configurator/src/main/java/org/apereo/openequella/adminconsole/configurator/Configurator.java

```diff
@@ -0,0 +1,288 @@
+        if (source == addServerButton) {
+            write = handleAddServer();
+        } else if (source == editServerButton) {
+            write = handleEditServer(index);
+        } else if (source == removeServerButton) {
+            write = handleRemoveServer(index);
+        } else if (source == editProxyButton) {
+            write = handleEditProxy();
+        }
+
+        if (write) {
+            config.writeServerConfigFile();
+        }
+    }
+
+    private void updateButtons() {
+        boolean enable = (serverPicker.getSelectedIndex() > -1);
+        editServerButton.setEnabled(enable);
+        removeServerButton.setEnabled(enable);
+    }
+
+    protected boolean handleAddServer() {
+        final ServerProfile selectedProfile = (ServerProfile) serverPicker.getSelectedItem();
+        final ServerProfile newProfile = new ServerProfile();
+        if (selectedProfile != null
+                && JOptionPane.showConfirmDialog(this, "Do you want to clone the currently selected server?", "Clone?",
+                        JOptionPane.YES_NO_OPTION) == JOptionPane.YES_OPTION) {
+            newProfile.setName(selectedProfile.getName());
+            newProfile.setUrl(selectedProfile.getUrl());
+        }
+
+        final AppServerEditDialog dialog = new AppServerEditDialog(this, newProfile);
+        dialog.setModal(true);
+        dialog.setVisible(true);
+
+        if (dialog.getResult() == AppServerEditDialog.RESULT_OK) {
+            final ServerProfile profile = dialog.getUpdatedProfile();
+            config.getServers().add(profile);
+
+            serverPicker.addItem(profile);
+            serverPicker.setSelectedIndex(serverPicker.getItemCount() - 1);
+            updateButtons();
+            return true;
+        }
+        return false;
+    }
+
+    protected boolean handleEditServer(int index) {
+        if (index > -1) {
+            final ServerProfile profile = (ServerProfile) serverPicker.getSelectedItem();
+            AppServerEditDialog dialog = new AppServerEditDialog(this, profile);
+            dialog.setModal(true);
+            dialog.setVisible(true);
+
+            if (dialog.getResult() == AppServerEditDialog.RESULT_OK) {
+
+                final ServerProfile editedProfile = dialog.getUpdatedProfile();
+                serverPicker.removeItemAt(index);
+                serverPicker.insertItemAt(editedProfile, index);
+                serverPicker.setSelectedIndex(index);
+
+                updateButtons();
+                return true;
+            }
+        }
+        return false;
+    }
+
+    protected boolean handleRemoveServer(int index) {
+        if (index > -1) {
+            final ServerProfile profile = (ServerProfile) serverPicker.getSelectedItem();
+            final int result = JOptionPane.showConfirmDialog(this, "Are you sure you want to remove this server?",
+                    "Are you sure?", JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
```

Comment:
Just remind that these strings need to be declared as constants :)

Reply:
Not sure it was a big deal, but I have done so :)